File: tools/parity_qwenimage_vae.jl
# Qwen-Image 2.1 VAE decoder: this tree against PyTorch, on latents the model
# actually produces.
#
#     julia --project=<repo-root> tools/parity_qwenimage_vae.jl [assetdir]
#     benchmark/torch.sh tools/parity_qwenimage_vae.py
#
# Two processes and two files between them, the same shape as
# `gap_vs_rocm.jl`/`.py`: they need different runtimes, and the INPUT is stated
# once — here — so both sides decode the identical latents.
#
# ## Latents the model actually produces
#
# The latents come out of twenty real denoise steps rather than from the
# exporter's `vae_reference.safetensors`, whose `randn` latents are far outside
# anything the denoiser emits. Until 2026-09-26 that reference could not be
# reproduced at all: the VAE was exported in fp16, its intermediates left fp16's
# 65504 range and 81% of the output came back NaN — the same failure that turned
# a transparent background into NaN. The VAE ships in fp32 now and matches that
# reference too (mean |diff| 1.1e-6), but real latents remain the fairer test.
#
# Measured 2026-09-26 at 1024², the fp32 decoder against diffusers at fp32 on the
# latents of a transparent-background prompt: mean |alpha diff| 1.4e-6, max
# 8.5e-5. The fp16 decoder it replaced was at 1.8e-3 against a bf16 reference.

using QwenImageRunner, DNNKernels, Mantle, KernelAbstractions, Random, Printf
const QIR = QwenImageRunner
const KA = KernelAbstractions

const PROMPT = "a red apple on a wooden table, studio photo"
const W = 256
const H = 256
const STEPS = 20
const SEED = 7

dir = isempty(ARGS) ? QIR.assetdir() : ARGS[1]
out = joinpath(@__DIR__, "..", "tmp")
mkpath(out)
backend = Mantle.LavaBackend()

lw, lh = W ÷ 16, H ÷ 16
@printf("prompt %s\n  %dx%d, latent %dx%d, %d steps, graph from %s\n",
        repr(PROMPT), W, H, lh, lw, STEPS, dir)

encoder = QIR.qwenimagetextencoder(; backend)
emb = QIR.encode_prompt(encoder, PROMPT)
Mantle.release!(encoder)
encoder = nothing
GC.gc(true); Mantle.trim_gpu_pool!()

transformer = QIR.qwenimagetransformer(; context_tokens = size(emb, 2), dir, latent = (lh, lw))
schedule = QIR.qwen_schedule(W, H; steps = STEPS)
Random.seed!(SEED)
latents = DNNKernels.toback(backend, Float16.(randn(Float32, 64, lh * lw, 1)))
timestep = similar(latents, eltype(latents), size(latents, 3))
embeddings = DNNKernels.toback(backend, emb)
for i in eachindex(schedule.timesteps)
    fill!(timestep, convert(eltype(timestep), schedule.timesteps[i] / 1000.0f0))
    prediction = QIR.denoise!(transformer, latents, embeddings, timestep)
    DNNKernels.eulerstep!(latents, prediction, schedule.sigmas[i], schedule.sigmas[i + 1])
end
KA.synchronize(backend)
Mantle.release!(transformer)
transformer = nothing
GC.gc(true); Mantle.trim_gpu_pool!()

grid = QIR.unpacklatents(Array(latents), lw, lh)
grid5 = reshape(grid, lw, lh, 1, size(grid, 3), size(grid, 4))

vae = QIR.qwenimagevae(; backend, dir)
image = Float32.(Array(QIR.decode!(vae, DNNKernels.toback(backend, Float16.(grid5)))))
KA.synchronize(backend)

# Raw Float32, no header. A Julia array of `(w, h, 1, c, b)` in column-major
# order is byte-identical to a C-order `(b, c, 1, h, w)` — torch's layout — so
# the reader needs no transpose and there is no axis convention to get wrong.
write(joinpath(out, "parity_latents.bin"), Float32.(vec(grid5)))
write(joinpath(out, "parity_ours.bin"), Float32.(vec(image)))
@printf("latents %s, image %s, %d NaN\n  wrote tmp/parity_latents.bin and tmp/parity_ours.bin\n",
        size(grid5), size(image), count(isnan, image))
@printf("now: benchmark/torch.sh tools/parity_qwenimage_vae.py\n")

File: tools/parity_qwenimage_vae.py
"""Qwen-Image 2.1 VAE decoder: PyTorch's answer to the latents Julia just made.

    julia --project=<repo-root> tools/parity_qwenimage_vae.jl
    benchmark/torch.sh tools/parity_qwenimage_vae.py

The second half of `parity_qwenimage_vae.jl`; that file's header says why the
latents come from twenty real denoise steps rather than from the exporter's own
`vae_reference.safetensors`, and carries the measured numbers.

Writes `tmp/parity_torch.bin` so the Julia side can render the two against each
other.
"""

import os, sys
import numpy as np, torch

# After `import torch`, exactly as benchmark/torch_models.py does it: the ROCm
# interpreter has torch and little else, and the project .venv that supplies
# safetensors/diffusers also contains a CUDA torch that must not shadow it.
for _p in filter(None, os.environ.get("JULIAVISION_TORCH_EXTRA_SITE", "").split(os.pathsep)):
    if _p not in sys.path:
        sys.path.append(_p)

sys.path.insert(0, "tools")
from artifacts import artifact
from export_qwenimage21 import import_qwen21, NormalizedVAEDecoder

lw = lh = 16
# Julia wrote (w,h,1,c,b) column-major, which is byte-identical to C-order
# (b,c,1,h,w) — the torch layout — so this needs no transpose.
lat = np.fromfile("tmp/parity_latents.bin", dtype=np.float32).reshape(1, 64, 1, lh, lw)
ours = np.fromfile("tmp/parity_ours.bin", dtype=np.float32).reshape(1, 4, 1, 256, 256)[:, :, 0]

_transformer_module, module = import_qwen21(artifact("diffusers-src") / "src")
vae = module.AutoencoderKLQwenImage21.from_pretrained(
    "Qwen/Qwen-Image-2.1", subfolder="vae", torch_dtype=torch.bfloat16,
    low_cpu_mem_usage=True).eval()
with torch.no_grad():
    ref = NormalizedVAEDecoder(vae).eval()(torch.from_numpy(lat).to(torch.bfloat16)).float().numpy()

d = np.abs(ours - ref)
sc = float(np.abs(ref).max())
print(f"shapes  ours {ours.shape}  torch {ref.shape}")
print(f"NaN     ours {int(np.isnan(ours).sum())}   torch {int(np.isnan(ref).sum())}")
print(f"max|d|  {d.max():.5f}   ({100*d.max()/sc:.3f}% of the reference's {sc:.4f} range)")
print(f"mean|d| {d.mean():.6f}      RMS {np.sqrt((d**2).mean()):.6f}")
x, y = ours.ravel() - ours.mean(), ref.ravel() - ref.mean()
print(f"corr    {float((x*y).sum()/np.sqrt((x*x).sum()*(y*y).sum())):.7f}")
for c in range(4):
    dc = d[0, c]
    print(f"  ch{c}   max {dc.max():.5f}  mean {dc.mean():.6f}")
np.fromfile  # keep ruff quiet
ref.astype(np.float32).tofile("tmp/parity_torch.bin")
